source/yats/include/yats/condition.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <condition_variable>
+#include <iostream>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#include <yats/constraint.h>
+
+namespace yats
+{
+
+class condition
+{
+public:
+    condition(size_t number_of_threads, size_t number_of_constraints)
+        : m_number_of_threads(number_of_threads)
+        , m_thread_identifier(number_of_constraints)
+        , m_is_active(true)
+        , m_task_added(number_of_constraints + 1)
+        , m_notify_count(number_of_constraints + 1, 0)
+    {
+        m_notify_count[m_thread_identifier] = m_number_of_threads;
+    }
+
+    class thread_guard
+    {
+    public:
+        thread_guard(condition* condition, bool is_active, std::unique_lock<std::mutex>& guard)
+            : m_condition(condition)
+            , m_is_active(is_active)
+        {
+            // Wait on the resource condition.
+            while (condition->m_notify_count[m_condition->m_thread_identifier] == 0 && m_is_active)
+            {
+                condition->m_task_added[m_condition->m_thread_identifier].wait(guard);
+            }
+            --condition->m_notify_count[m_condition->m_thread_identifier];
+        }
+
+        ~thread_guard()
```

Comment:
rule of three.. we should probably explicitly make this non-copyable, even if this is just a nested class